crates/circuit/src/register_data.rs

```diff
@@ -0,0 +1,262 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2025
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use std::sync::OnceLock;
+use std::{fmt::Debug, marker::PhantomData};
+
+use hashbrown::HashMap;
+use pyo3::{
+    prelude::*,
+    types::{IntoPyDict, PyDict, PyList},
+};
+
+use crate::{circuit_data::CircuitError, register::Register};
+
+/// Represents the location in which the register is stored within [RegisterData].
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct RegisterIndex<R: Register> {
+    index: u32,
+    _marker: PhantomData<R>,
+}
+
+impl<R: Register + Clone> Copy for RegisterIndex<R> {}
```

Comment:
You can't just derive `Copy` here, I would have expected it to just work without needing to put an explicit marker impl like this?

Reply:
`PhantomData` is weird about derives sometimes.